### ISSUE TITLE
[Roadmap]Prefill Context Parallel Refactor

### ISSUE THREAD

**Issue by Fridge003** [sgl-project/sglang#27252]
**This refactor design is generated by Codex under supervision. Technical details are prone to changes, but the overall design will not change.**

## Motivation

The current context parallelism (CP) implementation has grown around model- and
backend-specific branches. This makes the behavior hard to reason about and hard
to extend.

Main issues:

- Server args are confusing. Generic CP, DSA CP, and deprecated NSA aliases use
  separate flags and mode fields, even though they describe the same two CP
  sharding strategies.
- CP sharding logic is coupled with model code and attention backend code.
  Zigzag (`in-seq-split`) and interleave (`round-robin-split`) are model-agnostic
  layouts, but today each model family wires them differently.
- Model coverage is asymmetric. DSA models support both zigzag and interleave;
  MLA and GQA models historically mostly supported zigzag; DeepSeek V4 uses
  interleave. The draft CP-v2 path now also validates Qwen3 interleave, which
  is a useful GQA/MoE example for the new abstraction.
- Legacy communication code is scattered. `communicator_dsa_cp.py`, CP branches
  in model files, DSA indexer logic, and attention backend logic all own pieces
  of the same layout transition.
- Some topology choices are hardcoded in server args, such as forcing
  `enable_dp_attention=True` / `dp_size=1` for interleave-style paths.

The goal is to make CP a strategy-owned runtime abstraction: server args select
whether CP is enabled and which strategy to use; models and attention backends
call strategy interfaces instead of open-coding CP-specific branches.

### 1. How Server Args Usage Changes

Before, CP had overlapping flags:

```text
--enable-prefill-context-parallel
--prefill-cp-mode {in-seq-split}
--enable-dsa-prefill-context-parallel
--dsa-prefill-cp-mode {in-seq-split,round-robin-split}
--enable-nsa-prefill-context-parallel
--nsa-prefill-cp-mode
--attn-cp-size N
```

After, CP has one enable flag and one strategy axis:

```text
--enable-prefill-cp
--cp-strategy {zigzag,interleave}
--attn-cp-size N
```

Deprecated mode values map to the new strategy names:

| Deprecated value | New strategy |
|---|---|
| `in-seq-split` | `zigzag` |
| `round-robin-split` | `interleave` |

Deprecated boolean flags are still accepted and mirrored into
`enable_prefill_cp`:

```text
--enable-prefill-context-parallel
--enable-dsa-prefill-context-parallel
--enable-nsa-prefill-context-parallel
```

The normalized server args initialize a process-wide CP strategy singleton.
The runtime execution path remains gated by `SGLANG_ENABLE_CP_V2` during the
transition.

### 2. CP Strategy Design

#27313 lands the strategy abstraction scaffolding and process-wide strategy
initialization. CP strategy classes own the common policy surface for a CP layout:

- whether CP can apply for a forward batch;
- how per-forward CP metadata is built;
- how hidden states and position IDs are sharded;
- how rank-local hidden states and KV payloads are gathered;
- how per-request sharding will be expressed by strategies that need it;
- how model inputs are split before `model.forward`;
- how attention dispatch, KV materialization, and attention metadata reindexing
  will be routed once the model/backend migrations land.

The module layout is:

```text
python/sglang/srt/layers/cp/
  __init__.py      # package exports, including singleton helpers from base.py
  base.py          # enums, base metadata, ContextParallelStrategy ABC, singleton helpers
  zigzag.py        # implemented former in-seq-split strategy and zigzag metadata
  interleave.py    # former round-robin-split strategy shell and interleave metadata
  utils.py         # public import facade for strategy classes and metadata
```

There is no separate `strategy.py` in this PR. The process-wide helpers live in
`base.py`; `utils.py` is only the public import facade for strategy classes and
metadata types.

Class definitions:

```python
class ContextParallelStrategy(ABC):
    name: str
    kind: ContextParallelStrategyKind
    cp_size: int
    cp_rank: int
    per_layer_attn_cp_comm: bool

    def can_apply(self, num_tokens, forward_batch) -> bool: ...
    def build_metadata(self, num_tokens, seqs_len, extend_seqs_len=None): ...
    def shard_hidden_states(self, x, forward_batch): ...
    def shard_position_ids(self, positions, forward_batch): ...
    def gather_hidden_states(self, x, forward_batch, stream=None): ...
    def gather_kv_cache(self, x, forward_batch, stream=None): ...
    def shard_per_request(self, extend_seqs_cpu, extend_seqs): ...
    def split_before_forward(self, forward_batch, input_ids, positions, input_embeds=None): ...
    def run_attention(self, q, forward_batch, device, attn_fn, attention_backend): ...
    def materialize_full_kv(self, forward_batch, layer, k, v): ...
    def reindex_attn_metadata(self, core_attn_metadata): ...

class ZigzagCPStrategy(ContextParallelStrategy): ...
class InterleaveCPStrategy(ContextParallelStrategy): ...
```

Process-wide helpers added by #27313:

```python
def init_cp_strategy(server_args) -> None: ...
def get_cp_strategy(): ...
def get_cp_strategy_kind() -> ContextParallelStrategyKind: ...
def is_cp_enabled() -> bool: ...
def is_zigzag() -> bool: ...
def is_interleave() -> bool: ...
```

`ServerArgs._handle_context_parallelism()` now calls `init_cp_strategy(self)`
after legacy CP arguments are normalized and validated. If `enable_prefill_cp` is
false, or `attn_cp_size <= 1`, the singleton is cleared. Otherwise,
`cp_strategy={zigzag,interleave}` binds either `ZigzagCPStrategy` or
`InterleaveCPStrategy`.

`get_cp_strategy()` also lazily initializes from global server args. This lets
worker processes recover the singleton after deserialization, where
`ServerArgs.__post_init__` is not rerun. The singleton and compatibility
predicates are intentionally visible under both CP-v1 and CP-v2. The
`SGLANG_ENABLE_CP_V2` env var is introduced as a default-off runtime rollout
flag for future CP-v2 dispatch, but #27313 does not make `get_cp_strategy()`
return `None` when that env var is false.

Metadata definitions:

- `BaseContextParallelMetadata` stores `total_seq_lens` and `bs`.
- `ZigzagContextParallelMetadata` extends it with the former in-seq-split
  permutation fields, per-rank token counts, per-sequence FlashAttention tensors,
  and CPU-side diagnostic lists. `ContextParallelMetadata` remains an alias to
  the zigzag metadata shape.
- `InterleaveContextParallelMetadata` extends the base metadata and currently has
  no extra per-forward payload.

Class relationship:

```mermaid
classDiagram
    class ContextParallelStrategyKind {
        <<enumeration>>
        NONE
        ZIGZAG
        INTERLEAVE
        from_string()
        cli_value()
    }

    class CPAttentionBackendKind {
        <<enumeration>>
        FLASH_ATTENTION
        from_string()
    }

    class BaseContextParallelMetadata {
        total_seq_lens: int
        bs: int
    }

    class ZigzagContextParallelMetadata {
        split_list
        zigzag_index
        cp_reverse_index
        reverse_split_len
        per_rank_actual_token
        max_rank_len
        kv_len_prev_tensor
        kv_len_next_tensor
        actual_seq_q_prev_tensor
        actual_seq_q_next_tensor
        cu_seqlens_q_prev_tensor
        cu_seqlens_q_next_tensor
        total_q_prev_tokens
        total_q_next_tokens
        max_seqlen_q_prev
        max_seqlen_q_next
        kv_len_prev_list
        kv_len_next_list
        actual_seq_q_prev_list
        actual_seq_q_next_list
    }

    class InterleaveContextParallelMetadata {
        no extra payload yet
    }

    class ContextParallelStrategy {
        <<abstract>>
        name: str
        kind: ContextParallelStrategyKind
        cp_size: int
        cp_rank: int
        per_layer_attn_cp_comm: bool
        can_apply()
        build_metadata()
        shard_hidden_states()
        shard_position_ids()
        gather_hidden_states()
        gather_kv_cache()
        shard_per_request()
        split_before_forward()
        run_attention()
        materialize_full_kv()
        reindex_attn_metadata()
    }

    class ZigzagCPStrategy {
        kind = ZIGZAG
    }

    class InterleaveCPStrategy {
        kind = INTERLEAVE
    }

    BaseContextParallelMetadata <|-- ZigzagContextParallelMetadata
    BaseContextParallelMetadata <|-- InterleaveContextParallelMetadata
    ContextParallelStrategy <|-- ZigzagCPStrategy
    ContextParallelStrategy <|-- InterleaveCPStrategy
    ContextParallelStrategy --> ContextParallelStrategyKind
    ContextParallelStrategy --> CPAttentionBackendKind
    ZigzagCPStrategy --> ZigzagContextParallelMetadata
    InterleaveCPStrategy --> InterleaveContextParallelMetadata
```

Initialization flow added by #27313:

```mermaid
flowchart TD
    A["ServerArgs._handle_legacy_cp_arguments()"] --> B["ServerArgs._handle_context_parallelism()"]
    B --> C["validate canonical CP config"]
    C --> D["init_cp_strategy(self)"]
    D --> E["base._STRATEGY"]
    E --> F["get_cp_strategy()"]
    E --> G["get_cp_strategy_kind() / is_cp_enabled() / is_zigzag() / is_interleave()"]
    F --> H["future CP-v2 model, attention, and communicator dispatch"]
```

The two strategies differ only in layout policy at this stage:

- `ZigzagCPStrategy` represents the former `in-seq-split` layout. It splits each
  sequence into `2 * cp_size` blocks and gives each rank one early block and one
  late block. This balances causal attention cost. Its shell currently builds
  base zigzag metadata and leaves concrete sharding, gather, attention dispatch,
  and KV materialization to follow-up PRs.
- `InterleaveCPStrategy` represents the former `round-robin-split` layout. It
  assigns token `i` to rank `i % cp_size`. Its shell currently builds base
  interleave metadata and leaves concrete sharding, gather, attention dispatch,
  and KV materialization to follow-up PRs.

### 3. Model Runner CP-v2 Boundary

#28421 keeps CP-v2 sequence sharding and final gather at the runner boundary
instead of embedding that orchestration in Qwen model files. The original merge
wired this path in `model_runner.py`; after the later runner refactor, the eager
extend implementation lives in
`python/sglang/srt/model_executor/runner/eager_runner.py`, while
`model_runner.py` remains the top-level dispatcher and avoids prefill CUDA graph
replay when a CP strategy is configured.

For an active CP-v2 extend batch, the eager runner now:

1. calls `is_cp_v2_active(forward_batch)` after forward metadata initialization;
2. calls `prepare_cp_forward(forward_batch)` to build
   `forward_batch.attn_cp_metadata`;
3. materializes complete input embeddings when `input_embeds` is absent;
4. calls `cp_split_before_forward(...)` to produce rank-local hidden states and
   position IDs;
5. runs `model_runner.model.model(...)` directly with sharded embeddings and
   positions, bypassing the model wrapper's legacy CP/logits path;
6. on the final PP rank, calls `cp_gather_after_forward(...)` before invoking
   the logits processor.

This makes the model boundary responsible for only two layout transitions:

- full input embeddings/positions -> rank-local CP layout before the model body;
- rank-local hidden states -> full token order before logits.

The attention backend owns the per-layer CP-v2 attention transitions. In
FlashAttention, when `is_cp_v2_active(forward_batch)` is true:

- KV cache writes call `cp_strategy.materialize_full_kv(...)`;
- attention calls route through `cp_strategy.run_attention(...)`.

### 4. How To Handle Communication

Communication should be owned by the layer that naturally owns the layout
transition:

- Interconnect communication between CP and EP/DP/TP, such as gather/scatter
  around MLP, is handled by `communicator.py`.
- AllGather before attention kernels is handled by CP strategies through
  `gather_kv_cache` / `materialize_full_kv`.
- Scatter/gather at model entry and exit is handled by CP strategies through
  `split_before_forward` and gather helpers.
- `communicator_dsa_cp.py` should be removed. Communication operations should
  not be exposed directly in modeling files such as `dsa_indexer.py` or
  `deepseek_v4.py`.

The base communicator owns CP `SCATTERED`/`FULL` hidden-state transitions:

```mermaid
flowchart TD
    L["LayerCommunicator.__init__"] --> C["CommunicateContext.init_new()"]
    C --> Gate["_is_attn_cp_layer_comm_enabled()"]
    Gate -->|false| Normal["normal communicator dispatch tables"]
    Gate -->|true| CPTable["CP dispatch table"]

    CPTable --> A["prepare_attn: SCATTERED -> SCATTERED"]
    CPTable --> B["prepare_mlp: SCATTERED -> FULL or SCATTERED"]
    CPTable --> D["postprocess_layer: FULL or SCATTERED -> SCATTERED"]

    B --> Gather["_cp_gather_hidden_states_and_residual"]
    D --> Scatter["_cp_scatter_hidden_states"]
    Gather --> AG["attn_cp_all_gather_into_tensor"]
    Scatter --> RS["attn_cp_reduce_scatter_tensor"]
```

Exact layout dispatch should run before same-group-size shortcuts, because under
CP two tensors can have the same process group size while still representing
different token layouts.

### 5. Backward Compatibility Design

The refactor should merge gradually and stay off by default until the new path
has enough coverage.

CP-v2 runtime dispatch is controlled by `SGLANG_ENABLE_CP_V2`. 

`enable_cp_v2()` reads `SGLANG_ENABLE_CP_V2` through
`sglang.srt.environ.envs`. `is_cp_v2_active(forward_batch)` is the runtime
dispatch predicate for per-batch CP-v2 execution; it additionally requires
context-parallel extend mode, an initialized strategy, `input_ids`, and
`strategy.can_apply(...)`.

When `SGLANG_ENABLE_CP_V2=0`, server args still normalize the CP configuration
and initialize the strategy singleton for compatibility predicates. Runtime
dispatch should use `is_cp_v2_active(forward_batch)`, not `get_cp_strategy()`
alone. When `SGLANG_ENABLE_CP_V2=1`, `is_cp_v2_active(forward_batch)` exposes
active batches to model runner, attention backends, model bodies, and
communication code. 

Compatibility details:

- Deprecated CLI flags are still parsed and warn.
- Deprecated mode flags also enable `enable_prefill_cp`.
- `ContextParallelMetadata` remains importable as an alias to the zigzag
  metadata shape.
- `cp/utils.py` exposes CP-v2 helper names such as `enable_cp_v2`,
  `is_cp_v2_active`, `prepare_cp_forward`, `cp_split_before_forward`, and
  `cp_gather_after_forward`.
- Non-DSA deprecated predicates preserve their old server-args semantics.
- DSA deprecated predicates remain in `attention/dsa/utils.py` while the CP-v2
  runtime helpers live in `cp/utils.py`.

## Roadmap
An example PR for this refactor effort can be found at #25957. It will be split into a series of smaller PRs as the following plan:
- [x] Simplify the server args usage with the new ones #27312 
- [x] Add CP strategy abstractions. 
    - [x] defined classes, metadata, add related unit tests, and add the flag `SGLANG_ENABLE_CP_V2` for backward compatibility. #27313
    - [x] Implement member functions of zigzag strategy #28421
    - [x] Implement member functions of interleave strategy #30482
- [ ] Move communication ownership into the common communicator. #28775
  - Fold DSA CP `SCATTERED`/`FULL` transitions into `communicator.py`.
  - Remove `communicator_dsa_cp.py`.
  - Keep model files focused on model computation rather than layout
        communication.
- [x] Implement the new CP design on covered model families
  - [x] Support GQA models, starting with Qwen3, also include GPT-OSS/Mimo V2.5 #29972 
  - [x] Support MLA models, including Kimi K2.5 and DeepSeek V3 #31619 
  - [x] Support DSA models, including GLM-5 and DeepSeek-V3.2 #30482
  - [x] Support DeepSeek V4 with the new CP implementation #33532
- [ ] Enhance the Context Parallel tests with long context pressure tests.
- [ ] Flip `SGLANG_ENABLE_CP_V2`  to True and deprecate old CP implementation
  - [x] Deprecate CP V1 tests https://github.com/sgl-project/sglang/pull/36222
  - [ ]  Remove server args/environ settings related to CP V1 https://github.com/sgl-project/sglang/pull/36223/
  - [ ]  Remove old logics in CP V1 logics https://github.com/sgl-project/sglang/pull/36228
  - [ ] Canonicalize prefill CP API names #36229 
  - [ ] Update documentation for prefill CP V2 https://github.com/sgl-project/sglang/pull/36230
- [ ] Future enhancement, such as supporting more attention backends with CP, or adding documents
